cmd/metal-api/internal/datastore/event.go

```diff
@@ -62,7 +79,121 @@ func (rs *RethinkStore) ProvisioningEventForMachine(log *zap.SugaredLogger, even
 	}
 
 	newEC.TrimEvents(100)
-
 	err = rs.UpsertProvisioningEventContainer(newEC)
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Message == states.FSMMessageEnterWaitingState || config.Message == states.FSMMessageLeaveWaitingState {
+		err = rs.AdjustNumberOfWaitingMachines(log, publisher, machine)
+	}
+
 	return newEC, err
 }
+
+func (rs *RethinkStore) AdjustNumberOfWaitingMachines(log *zap.SugaredLogger, publisher bus.Publisher, machine *metal.Machine) error {
+	p, err := rs.FindPartition(machine.PartitionID)
+	if err != nil {
+		return err
+	}
+
+	waitingMachines, err := rs.listWaitingMachines(machine.PartitionID, machine.SizeID)
+	if err != nil {
+		return err
+	}
+
+	poolSizeExceeded := 0
+	if strings.Contains(p.WaitingPoolSize, "%") {
+		poolSizeExceeded = rs.waitingMachinesMatchPoolSize(len(waitingMachines), p.WaitingPoolSize, true, machine)
+	} else {
+		poolSizeExceeded = rs.waitingMachinesMatchPoolSize(len(waitingMachines), p.WaitingPoolSize, false, machine)
+	}
+
+	if poolSizeExceeded == 0 {
+		return nil
+	}
+
+	if poolSizeExceeded > 0 {
+		m := randomMachine(waitingMachines)
+		err = rs.machineCommand(log, &m, publisher, metal.MachineOffCmd, metal.ShutdownState)
+		if err != nil {
+			return err
+		}
+	} else {
+		shutdownMachines, err := rs.listShutdownMachines(machine.PartitionID, machine.SizeID)
+		if err != nil {
+			return err
+		}
+
+		m := randomMachine(shutdownMachines)
+		err = rs.machineCommand(log, &m, publisher, metal.MachineOnCmd, metal.AvailableState)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (rs *RethinkStore) waitingMachinesMatchPoolSize(actual int, required string, inPercent bool, m *metal.Machine) int {
+	if !inPercent {
+		r, err := strconv.ParseInt(required, 10, 32)
+		if err != nil {
+			return 0
+		}
+		return actual - int(r)
+	}
+
+	p := strings.Split(required, "%")[0]
+	r, err := strconv.ParseInt(p, 10, 32)
+	if err != nil {
+		return 0
+	}
+
+	allMachines, err := rs.listMachinesInPartition(m.PartitionID, m.SizeID)
+	if err != nil {
+		return 0
+	}
+
+	return actual - int(math.Round(float64(len(allMachines))*float64(r)/100))
+}
+
+func randomMachine(machines metal.Machines) metal.Machine {
+	var idx int
+	b, err := rand.Int(rand.Reader, big.NewInt(int64(len(machines)))) //nolint:gosec
+	if err != nil {
+		idx = int(b.Uint64())
+	} else {
+		mathrand.Seed(time.Now().UnixNano())
+		// nolint
+		idx = mathrand.Intn(len(machines))
+	}
+
+	return machines[idx]
+}
+
+func (rs *RethinkStore) machineCommand(logger *zap.SugaredLogger, m *metal.Machine, publisher bus.Publisher, cmd metal.MachineCommand, state metal.MState) error {
+	evt := metal.MachineEvent{
+		Type: metal.COMMAND,
+		Cmd: &metal.MachineExecCommand{
+			Command:         cmd,
+			TargetMachineID: m.ID,
+			IPMI:            &m.IPMI,
+		},
+	}
+
+	newMachine := *m
+	newMachine.State = metal.MachineState{Value: state}
```

Comment:
A proper description should be provided in order to see that the scaling algorithm put the machine into this state.